accounts/abi/abigen/bindv2.go

```diff
@@ -0,0 +1,361 @@
+		if args[i].Name == "" {
+			args[i].Name = fmt.Sprintf("Arg%d", i)
+		}
+		for index := 0; ; index++ {
+			if !used[args[i].Name] {
+				used[args[i].Name] = true
+				break
+			}
+			args[i].Name = fmt.Sprintf("%s%d", args[i].Name, index)
+		}
+	}
+	return args
+}
+
+// normalizeErrorOrEventFields normalizes errors/events for emitting through bindings:
+// Any anonymous fields are given generated names.
+func (cb *contractBinder) normalizeErrorOrEventFields(originalInputs abi.Arguments) abi.Arguments {
+	normalizedArguments := normalizeArgs(originalInputs)
+	for _, input := range normalizedArguments {
+		if hasStruct(input.Type) {
+			cb.binder.BindStructType(input.Type)
+		}
+	}
+	return normalizedArguments
+}
+
+// bindEvent normalizes an event and registers it to be emitted in the bindings.
+func (cb *contractBinder) bindEvent(original abi.Event) error {
+	// Skip anonymous events as they don't support explicit filtering
+	if original.Anonymous {
+		return nil
+	}
+	normalizedName, err := cb.registerIdentifier(cb.eventIdentifiers, original.Name)
+	if err != nil {
+		return err
+	}
+
+	normalized := original
+	normalized.Name = normalizedName
+	normalized.Inputs = cb.normalizeErrorOrEventFields(original.Inputs)
+	cb.events[original.Name] = &tmplEvent{Original: original, Normalized: normalized}
+	return nil
+}
+
+// bindError normalizes an error and registers it to be emitted in the bindings.
+func (cb *contractBinder) bindError(original abi.Error) error {
+	normalizedName, err := cb.registerIdentifier(cb.errorIdentifiers, original.Name)
+	if err != nil {
+		return err
+	}
+
+	normalized := original
+	normalized.Name = normalizedName
+	normalized.Inputs = cb.normalizeErrorOrEventFields(original.Inputs)
+	cb.errors[original.Name] = &tmplError{Original: original, Normalized: normalized}
+	return nil
+}
+
+func parseLibraryDeps(unlinkedCode string) (res []string) {
+	reMatchSpecificPattern, err := regexp.Compile(`__\$([a-f0-9]+)\$__`)
+	if err != nil {
+		panic(err)
+	}
+	for _, match := range reMatchSpecificPattern.FindAllStringSubmatch(unlinkedCode, -1) {
+		res = append(res, match[1])
+	}
+	return res
+}
+
+// iterSorted iterates the map in the lexicographic order of the keys calling
+// onItem on each. If the callback returns an error, iteration is halted and
+// the error is returned from iterSorted.
+func iterSorted[V any](inp map[string]V, onItem func(string, V) error) error {
+	var sortedKeys []string
+	for key := range inp {
+		sortedKeys = append(sortedKeys, key)
+	}
+	sort.Strings(sortedKeys)
+
+	for _, key := range sortedKeys {
+		if err := onItem(key, inp[key]); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// BindV2 generates a Go wrapper around a contract ABI. This wrapper isn't meant
+// to be used as is in client code, but rather as an intermediate struct which
+// enforces compile time type safety and naming convention as opposed to having to
+// manually maintain hard coded strings that break on runtime.
```

Comment:
`This wrapper isn't meant to be used as is in client code, but rather as an intermediate struct which enforces compile time type safety and naming convention`:  This doesn't make sense to me, and I'm trying to think about a better description.